include/inviwo/core/processors/poolprocessor.h

```diff
@@ -0,0 +1,518 @@
+                   std::is_invocable<Job, Progress>, std::is_invocable<Job, Stop, Progress>,
+                   std::is_invocable<Job, Progress, Stop>>;
+
+    using ValidArgs = std::tuple<std::tuple<>, std::tuple<Stop>, std::tuple<Progress>,
+                                 std::tuple<Stop, Progress>, std::tuple<Progress, Stop>>;
+
+    static constexpr size_t argIndex = util::index_of_derived<std::true_type, CallTest>();
+    using Args = std::tuple_element_t<argIndex, ValidArgs>;
+    using Result = decltype(std::apply(std::declval<Job>(), std::declval<Args>()));
+};
+
+}  // namespace pool::detail
+
+template <typename Result, typename Done>
+inline void PoolProcessor::callDone(
+    InviwoApplication* app, std::shared_ptr<pool::detail::StateTemplate<Result, Done>> state) {
+    static const auto done = [](PoolProcessor& p, auto state) {
+        try {
+            if constexpr (std::is_invocable_v<Done, Result>) {
+                state->done(state->futures.front().get());
+            } else {
+                std::vector<Result> results;
+                for (auto& res : state->futures) {
+                    results.push_back(res.get());
+                }
+                state->done(results);
+            }
+        } catch (...) {
+            p.handleError();
+        }
+    };
+
+    if (state->count.fetch_sub(1) == 1) {
+        app->dispatchFrontAndForget([state]() {
+            if (auto wrapper = state->processor.lock()) {
+                auto& p = wrapper->processor;
+                const bool isLast = p.removeState(state);
+                p.getProgressBar().setActive(p.hasJobs());
+                if (isLast) p.getProgressBar().hide();
+
+                if (p.queuedDispatch() && !p.queue_.empty()) {
+                    p.submit(p.queue_.front());
+                    p.queue_.clear();
+                    return;
+                }
+
+                if (state->stop) return;
+
+                if (isLast || p.keepOldJobs()) {
+                    done(p, state);
+                }
+            }
+        });
+    }
+}
+
+template <typename Job, typename Done>
+void PoolProcessor::dispatchMany(std::vector<Job> jobs, Done&& done) {
+    using Result = typename pool::detail::JobTraits<Job>::Result;
+
+    static_assert(std::is_invocable_v<Done, std::vector<Result>>);
+
+    if (!keepOldJobs()) stopJobs();
+
+    auto state = makeState<Result, Done>(jobs.size(), std::forward<Done>(done));
+    Submission sub{state, {}, [this]() { setupProgress<Job>(); }};
+
+    auto app = getNetwork()->getApplication();
+    size_t i = 0;
+    for (auto& job : jobs) {
+        auto task = makeTask<Result>(std::move(job), state->getStop(), state->getProgress(i++));
+        state->futures.push_back(task->get_future());
+        sub.tasks.emplace_back([state, task, app]() {
+            if (!state->stop) (*task)();
+            callDone(app, state);
+        });
+    }
+
+    if (delayDispatch()) {
+        queue_.clear();
```

Comment:
Could this introduce racing conditions? if the "delayer" tries to dequeue the same time as we clear or push_back at the same time?

Reply:
we only use queue_ in the main thread